libraries/AP_Mount/AP_Mount_Tusuav.cpp

```diff
@@ -0,0 +1,1109 @@
+
+    // fill in packet
+    FrameType frame_to_send{
+        .flag={0x00},
+        data_len : sizeof(gimbal_ctrl_settings),
+        cmd : static_cast<uint8_t>(GimbalCmd::MasCmd_Gimbal),
+        sub_cmd : static_cast<uint8_t>(GimbalSubCmd::GimbalSubCmd_PodCtrlReg),
+        data : &gimbal_ctrl_settings.all
+    };
+    
+    // send packet to gimbal
+    return send_packet(frame_to_send);
+}
+
+// send target pitch and yaw angles to gimbal
+// fill the settings for choosing an earth-frame / body-frame before calling send_target_angles_rates function
+bool AP_Mount_Tusuav::send_target_angles_rates(MessageGimbalAngleRates mode)
+{
+    // gimbal does not support lock in angle control mode
+    if (!set_lock(false)) {
+        return false;
+    }
+
+    // fill in packet
+    FrameType frame_to_send{
+        .flag={0},
+        data_len : sizeof(mode),
+        cmd : static_cast<uint8_t>(GimbalCmd::MasCmd_Gimbal),
+        sub_cmd : static_cast<uint8_t>(GimbalSubCmd::GimbalSubCmd_PlaneControlPara),
+        data : (uint8_t*)&mode
+    };
+
+    // send targets to gimbal
+    return send_packet(frame_to_send);
+}
+
+// send reference of the joystick rocker for moving the gimbal or tracking window
+// when gimbal runs in tracking mode this command will move the tracking window for selecting the target
+// when in other mode(except angle mode) this command will set the rates of the yaw axis and pitch axis(±500 map to max/min rate speed)
+bool AP_Mount_Tusuav::send_stick_pitch_yaw(int16_t pitch,int16_t yaw)
+{
+    // gimbal does not support lock in angle control mode
+    if (!set_lock(false)) {
+        return false;
+    }
+
+    if (pitch > 500) {
+		pitch = 500;
+	}
+	else if (pitch < -500) {
+		pitch = -500;
+	}
+
+    if (yaw > 500) {
+    	yaw = 500;
+	}
+	else if (yaw < -500) {
+		yaw = -500;
+	}
+
+    int16_t value[2] = {pitch,yaw};
+    // fill in packet
+    FrameType frame_to_send{
+        .flag={0},
+        data_len : sizeof(value),
+        cmd : static_cast<uint8_t>(GimbalCmd::MasCmd_Gimbal),
+        sub_cmd : static_cast<uint8_t>(GimbalSubCmd::gimbal_single_cmd_stick_control),
+        data : (uint8_t*)&value[0]
+    };
+    debug("sticking %d %d", value[0], value[1]);
+    // send targets to gimbal
+    return send_packet(frame_to_send);
+}
+
+// send vehicle attitude and position to gimbal
+bool AP_Mount_Tusuav::send_m_ahrs()
+{
+    // get current location
+    Location loc;
+    int32_t alt_amsl_cm = 0;
+    if (!AP::ahrs().get_location(loc) || !loc.get_alt_cm(Location::AltFrame::ABSOLUTE, alt_amsl_cm)) {
+        debug("tus ahrs fault:%d %d %d",(int)loc.lat,(int)loc.lng,(int)loc.alt); 
+
+    }
+
+    // get vehicle velocity in m/s in earth frame
+    Vector3f vel_NED;
+    IGNORE_RETURN(AP::ahrs().get_velocity_NED(vel_NED));
+    Vector3f acc_NED;
+    acc_NED = AP::ahrs().get_accel_ef();
+    Vector3f gyro_NED;
+    gyro_NED = AP::ahrs().get_gyro();
+
+    uint32_t gps_last_time = AP::gps().last_message_time_ms();
```

Comment:
This is making the assumption that the location for the vehicle is coming from a GPS.

That may not be the case.

Reply:
Thank you for the suggestion! I have updated the code to use AHRS as the primary source for vehicle position, which should address the assumption of relying solely on GPS.

Reply:
Thank you for the suggestion! We’ve updated the code to use AP_HAL::millis() for the current timestamp, which should fix the issue.

Reply:
Thank you for your input. GPS is required in this case. Once a 2D fix is obtained, we increment the `_gps_update_count` and send it to the gimbal to ensure proper synchronization between the GPS data and other sensor data.

Reply:
Thank you for your input. GPS is required in this case. Once a 2D fix is obtained, we increment the '_gps_update_count' and send it to the gimbal to ensure proper synchronization between the GPS data and other sensor data.